lecture07/springboot-backend/src/main/java/net/javaguides/springboot/service/EmployeeService.java

```diff
@@ -0,0 +1,92 @@
+package net.javaguides.springboot.service;
+
+import net.javaguides.springboot.exception.ResourceNotFoundException;
+import net.javaguides.springboot.model.Employee;
+import net.javaguides.springboot.repository.EmployeeRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class EmployeeService {
+
+    /**
+     * コンストラクタインジェクション
+     * "@Autowired"は、コンストラクタが1件のみかつSpring4.3以上であれば省略可能だがつけておく
+     * フィールドにfinalを使用することができて不変性を担保できるため、推奨される
+     */
+    private final EmployeeRepository employeeRepository;
+    @Autowired
+    public EmployeeService(EmployeeRepository employeeRepository) {
+        this.employeeRepository = employeeRepository;
+    }
+
+    /**
+     * 全てのEmployeeデータを取得する
+     * @return 全てのEmployeeデータ
+     */
+    public List<Employee> getAllEmployee(){
+        return employeeRepository.findAll();
+    }
+
+    /**
+     * IDに紐づくEmployeeデータを取得する
+     * @param id ID
+     * @return IDに紐づくEmployeeデータ
+     */
+    public Employee getEmployeeById(long id){
+        Optional<Employee> employee = employeeRepository.findById(id);
+        return employee.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
+    }
+
+    /**
+     * Employeeデータを登録する
+     * @param employee 登録するEmployeeデータ
+     * @return Employeeデータ
+     */
+    public Employee registerEmployee(Employee employee) {
+        if(employeeRepository.countByEmailId(employee.getEmailId()) == 0){
+            // メールアドレスが未登録であれば登録する
+            return employeeRepository.save(employee);
+        }else{
+            // 登録済みであればエラーを返す
+            throw new DuplicateKeyException("Employee is registered email:" + employee.getEmailId());
+        }
+    }
+
+    /**
+     * Employeeデータを更新する
+     * @param id ID
+     * @param updateEmployee 更新するEmployeeデータ
+     * @return Employeeデータ
+     */
+    public Employee updateEmployee(long id, Employee updateEmployee) {
+        // idに紐づくEmployeeがなければコール側でエラー返却してくれる
+        Employee employee = getEmployeeById(id);
+        if(employeeRepository.countByEmailId(updateEmployee.getEmailId()) == 0) {
+            // 指定したIDをもつ かつ 同じメールアドレスが登録されていなければ、Employeeデータがあれば更新する
+            updateEmployee.setId(id);
+            // 更新したEmployeeを返却
+            return employeeRepository.save(updateEmployee);
+        }else{
+            // 登録済みであればエラーを返す
+            throw new DuplicateKeyException("Employee is registered email:" + employee.getEmailId());
+        }
+    }
+
+    /**
+     * Employeeデータを削除する
+     * @param id ID
+     * @return Employeeデータ
+     */
+    public void deleteEmployee(long id) {
+        Employee employee = getEmployeeById(id);   // idに紐づくEmployeeがなければコール側でエラー返却してくれる
```

Comment:
これは気持ちはわかるのですが、Serviceのメソッド間での依存関係は持たせないほうがよいかと思います。
なぜならEmployeeをIDに応じて取得するのとIDで指定したEmployeeを削除するのは別の業務処理（ユースケースともいう）と捉えたいからです。

したがって下記のように実装するのが良いと私は考えます。

```suggestion
        Employee employee = employeeRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```

Reply:
同じような処理を複数記載するのはどうかと思い、上記のようにしたのですが、たしかに依存関係をもつことはよくないですね。
それなら、おっしゃられている通り別の業務処理として捉えた方がスマートだと思いました。
[20a7bc9](https://github.com/aoha968/Java2211/pull/8/commits/20a7bc9d214eb5aad7e73f5d958521fc3285bf6b)

Reply:
👍